File: 1_term_ICT/Programming/lectures/Demo/Demo_lec_04(22)/ООП/Customer1.py
class Customer:
    """Клиент телефонной компании."""

    def __init__(self, name, balance=0):
        self.name = name
        self._balance = balance

    def __str__(self):
        return "Клиент \"{}\". Баланс: {} руб.".format(self.name, self.balance)

    @property
    def balance(self):
        """Вернуть баланс клиента.
        Свойство 'balance' доступно только для чтения:
        """
        return self._balance

    def record_payment(self, amount_paid):
        """Пополнить баланс клиента на 'amount_paid' руб."""
        assert amount_paid > 0, "Сумма пополнения должна быть > 0!"
        self._balance += amount_paid

    def record_call(self, call_type, minutes):
        """Списать стоимость звонка с баланса клиента.
        Параметры:
            - call_type (str): тип звонка:
                "Г": городской;
                "М": мобильный;
            - minutes (float): количество минут разговора.
        """
        if call_type == "Г":
            self._balance -= minutes * 5
        elif call_type == "М":
            self._balance -= minutes * 1


if __name__ == "__main__":

    ivan = Customer("Иван Петров")
    elena = Customer("Елена Миронова", 100)

    ivan.record_call("Г", 20)
    ivan.record_call("М", 5)
    elena.record_call("М", 10)

    ivan.record_payment(155)    # Пополнили телефон на 155 руб.

    print(ivan)                 # Клиент "Иван Петров". Баланс: 50 руб.
    print(elena)                # Клиент "Елена Миронова". Баланс: 90 руб.

File: 2_term_ICT/Programming/lectures/Demo_OOP_C#/Customer1.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Customer
{
    class Customer
    {
        public string Name { get; set; }
        public double Balance { get; private set; }

        public Customer(string name, double balance = 100)
        {
            Name = name;
            Balance = balance;
        }

        public override string ToString() => $"Клиент: {Name} имеет баланс: {Balance}";
        
        public void RecordPayment(double amountPaid)
        {
            if (amountPaid > 0)
                Balance += amountPaid;
        }

        public void RecordCall(char callType, int minutes)
        {
            if (callType == 'Г')
                Balance -= minutes * 5;
            else
                if (callType == 'М')
                    Balance -= minutes * 1;
        }
    }

    class Customer1
    {
        static void Main(string[] args)
        {
            Customer Ivan = new Customer("Иван Петров", 500);
            Customer Elena = new Customer("Елена Иванова");
            Ivan.RecordCall('Г', 12);
            Elena.RecordCall('М', 25);

            Console.WriteLine(Ivan);
            Console.WriteLine(Elena);

        }
    }
}
